### PR TITLE
JENKINS-55920: add CasC support to GitHubPluginConfig w/Tests

### PR DESCRIPTION
**Pull request by er1c**
This is based upon both https://github.com/jenkinsci/github-plugin/pull/205 and https://github.com/jenkinsci/github-plugin/pull/209

I can rebase the PR once #209 is addressed to make the PR cleaner.

---
This change is [Reviewable](https://reviewable.io/reviews/jenkinsci/github-plugin/210)

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -182,20 +206,70 @@
...
+            <groupId>org.jenkins-ci.plugins</groupId>
+            <artifactId>matrix-auth</artifactId>
+            <version>2.3</version>
+            <scope>test</scope>
+        </dependency>
+
+        <dependency>
+            <groupId>io.jenkins</groupId>
+            <artifactId>configuration-as-code</artifactId>
+            <version>1.0</version>
```
Comment (timja) [inline pom.xml]: use the latest released version please (1.7) https://github.com/jenkinsci/configuration-as-code-plugin/releases

**Comment by casz** [comment]
@er1c would be great if you could rebase:
This diff is no good: https://github.com/dgarzon/github-plugin/compare/HOTFIX-add-casc-support...er1c:HOTFIX-add-casc-support

**Comment by er1c** [comment]
@casz  https://github.com/er1c/github-plugin/pull/1

File: pom.xml [hunk 2]
```diff
@@ -207,6 +207,13 @@
             <scope>test</scope>
         </dependency>
 
+        <dependency>
+            <groupId>io.jenkins</groupId>
+            <artifactId>configuration-as-code</artifactId>
+            <version>1.7</version>
```
Comment (casz) [inline pom.xml]: Please update to a newer version and consider adding the `tests` classifier.

File: src/test/java/org/jenkinsci/plugins/github/config/ConfigAsCodeTest.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+public class ConfigAsCodeTest {
+
+    @Rule
+    public JenkinsRule r = new JenkinsRule();
+
+    @Test
+    public void shouldSupportConfigurationAsCode() throws Exception {
+        ConfigurationAsCode.get().configure(
+                ConfigAsCodeTest.class.getResource("configuration-as-code.yml").toString());
```
Comment (casz) [inline src/test/java/org/jenkinsci/plugins/github/config/ConfigAsCodeTest.java]: once you add the `tests` classifier, please use the annotation `@ConfiguredWithCode("configuration-as-code.yml")`

**Comment by er1c** [comment]
This has been rebased, and the outstanding review comments addressed.

File: src/main/resources/org/jenkinsci/plugins/github/config/GitHubPluginConfig/config.groovy [hunk 4]
```diff
@@ -26,8 +26,7 @@ f.section(title: descriptor.displayName) {
                 table(width: "100%", style: "margin-left: 7px;") {
                     f.optionalBlock(title: _("Specify another hook URL for GitHub configuration"),
                             inline: true,
-                            field: "overrideHookUrl",
-                            checked: instance.overrideHookURL) {
+                            checked: instance.isOverrideHookUrl) {
```
Comment (KostyaSha) [inline src/main/resources/org/jenkinsci/plugins/github/config/GitHubPluginConfig/config.groovy]: either method call either method call via field name, i think it wouldn't work

Comment (casz) [reply src/main/resources/org/jenkinsci/plugins/github/config/GitHubPluginConfig/config.groovy]: JCasC does say no to optional blocks, use optional property https://github.com/jenkinsci/configuration-as-code-plugin/blob/master/docs/PLUGINS.md#step-4

Comment (casz) [reply src/main/resources/org/jenkinsci/plugins/github/config/GitHubPluginConfig/config.groovy]: Or move it out of the optional block, it smells :cry:

File: src/test/java/org/jenkinsci/plugins/github/config/ConfigAsCodeTest.java [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+                        .and(withCredsId(is("private_cred_id")))
+                        .and(withClientCacheSize(is(40)))
+                        .and(withIsManageHooks(is(false)))
+                ));
+    }
+
+    @Test
+    @ConfiguredWithCode("configuration-as-code.yml")
+    public void export_configuration() throws Exception {
+        /* TODO (From JCASC): Need to provide some YAML assertion library so that the resulting exported yaml
```
Comment (timja) [inline src/test/java/org/jenkinsci/plugins/github/config/ConfigAsCodeTest.java]: We need to update the plugin docs but it is being tested in this way in a few plugins and in the configuration-as-code repo:
https://github.com/jenkinsci/azure-keyvault-plugin/blob/master/src/test/java/org/jenkinsci/plugins/azurekeyvaultplugin/ConfigAsCodeTest.java#L33

File: src/test/resources/org/jenkinsci/plugins/github/config/configuration-as-code.yml [hunk 6]
```diff
@@ -0,0 +1,17 @@
+unclassified:
+
+  githubpluginconfig:
```
Comment (timja) [inline src/test/resources/org/jenkinsci/plugins/github/config/configuration-as-code.yml]: maybe another more minimal test which shows what you need if you set manage hooks false and no hook url?

File: src/main/resources/org/jenkinsci/plugins/github/config/GitHubPluginConfig/config.groovy (hunk 4 above)
Comment (er1c) [reply src/main/resources/org/jenkinsci/plugins/github/config/GitHubPluginConfig/config.groovy]: This isn't part of my wheel house, I think the way I have it setup, conforms to the example here: https://wiki.jenkins.io/display/JENKINS/Jelly+form+controls

I'm removing the "field" from the toggling of the checkbox to just toggle it based upon the value,  I'll see if I can add a test to verify that the checkbox is "checked" based upon the current code or correct if that doesn't work.

File: src/test/java/org/jenkinsci/plugins/github/config/ConfigAsCodeTest.java (hunk 5 above)
Comment (timja) [reply src/test/java/org/jenkinsci/plugins/github/config/ConfigAsCodeTest.java]: I've updated the docs now:
https://github.com/jenkinsci/configuration-as-code-plugin/blob/master/docs/PLUGINS.md#model-export-test

**Comment by startnow65** [comment]
@er1c Thanks for the good work!
Any idea when this would be merged ? I need this so I can set my hook url with JCasC.

Thanks

File: src/main/resources/org/jenkinsci/plugins/github/config/GitHubPluginConfig/config.groovy (hunk 4 above)
Comment (KostyaSha) [reply src/main/resources/org/jenkinsci/plugins/github/config/GitHubPluginConfig/config.groovy]: `isOverrideHookUrl()` or `instance.overrideHookURL`  code had correct call

File: src/test/java/org/jenkinsci/plugins/github/config/ConfigAsCodeTest.java [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+            stream can be checked for expected content. */
+        ConfigurationAsCode.get().export(System.out);
```
Comment (KostyaSha) [inline src/test/java/org/jenkinsci/plugins/github/config/ConfigAsCodeTest.java]: never user System.out, user loggers if you want to print something.

**Comment by m-barthelemy** [comment]
Hi, we're having this issue where `hookUrl` is ignored in configuration as code.
This PR seems to fix it. Will it get merged?

**Comment by oleg-nenashev** [comment]
I'd guess https://github.com/jenkinsci/github-plugin/pull/210/files#r277181851 is still a blocker, no?

**Comment by thirstydeveloper** [comment]
Any movement on this? My team is blocked by not being able to set the `overrideHookUrl` property.

**Comment by Misteur-Z** [comment]
Basically the only thing to set manually at startup or after a Jenkins upgrade. I would be happy to help with this issue, I can do some testing but I am no Java/Groovy dev.

**Comment by thefirstofthe300** [comment]
Based on @er1c's Github contribution history, it looks like he probably doesn't have bandwidth to work on this anymore. Does anyone here have the Java experience to take over this PR?

I am also hitting this particular issue.

**Comment by er1c** [comment]
> Based on @er1c's Github contribution history, it looks like he probably doesn't have bandwidth to work on this anymore. Does anyone here have the Java experience to take over this PR?
> 
> I am also hitting this particular issue.

I'm currently slammed and don't have time for this.  For anyone who is blocked, you can work around it via: https://gist.github.com/er1c/c057640d43c0dce149b0a0c23d1ce153#file-jenkins-github-yaml-L80

**Comment by casz** [comment]
Thanks @er1c for picking this up 👍

**Pull request closed without merging**